recipe-client-addon/lib/ShieldPreferences.jsm

```diff
@@ -0,0 +1,133 @@
+ */
+this.ShieldPreferences = {
+  async init() {
+    // Disabled outside of en-* locales temporarily (bug 1377192).
+    // Disabled when MOZ_DATA_REPORTING is false since the FHR UI is also hidden
+    // when data reporting is false.
+    if (AppConstants.MOZ_DATA_REPORTING && Services.locale.getAppLocaleAsLangTag().startsWith("en")) {
+      Services.obs.addObserver(this, "advanced-pane-loaded");
+      Services.prefs.addObserver(FHR_UPLOAD_ENABLED_PREF, this);
+
+      CleanupManager.addCleanupHandler(() => {
+        Services.obs.removeObserver(this, "advanced-pane-loaded");
+        Services.prefs.removeObserver(FHR_UPLOAD_ENABLED_PREF, this);
+      });
+    }
+  },
+
+  observe(subject, topic, data) {
+    switch (topic) {
+      // Add the opt-out-study checkbox to the Privacy preferences when it is shown.
+      case "advanced-pane-loaded":
+        if (!Services.prefs.getBoolPref("browser.preferences.useOldOrganization", false)) {
+          const checkbox = new this.OptOutStudyCheckbox(subject.document);
+          checkbox.inject();
+        }
+        break;
+      // If the FHR pref changes, set the opt-out-study pref to the value it is changing to.
+      case NS_PREFBRANCH_PREFCHANGE_TOPIC_ID:
+        if (data === FHR_UPLOAD_ENABLED_PREF) {
+          const fhrUploadEnabled = Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF);
+          Services.prefs.setBoolPref(OPT_OUT_STUDIES_ENABLED_PREF, fhrUploadEnabled);
+        }
+        break;
+    }
+  },
+
+  /**
+   * Injects the opt-out-study preference checkbox into about:preferences and
+   * handles events coming from the UI for it.
+   */
+  OptOutStudyCheckbox: class {
+    constructor(document) {
+      this.document = document;
+      this.refs = {};
+
+      const container = this.refs.container = document.createElementNS(XUL_NS, "vbox");
+      container.classList.add("indent");
+
+      const hContainer = this.refs.hContainer = document.createElementNS(XUL_NS, "hbox");
+      hContainer.setAttribute("align", "center");
+      container.appendChild(hContainer);
+
+      const checkbox = this.refs.checkbox = document.createElementNS(XUL_NS, "checkbox");
+      checkbox.setAttribute("id", "optOutStudiesEnabled");
+      checkbox.setAttribute("label", "Allow Firefox to install and run studies");
+      checkbox.setAttribute("preference", OPT_OUT_STUDIES_ENABLED_PREF);
+      checkbox.setAttribute("checked", Services.prefs.getBoolPref(OPT_OUT_STUDIES_ENABLED_PREF));
+      checkbox.setAttribute("disabled", !Services.prefs.getBoolPref(FHR_UPLOAD_ENABLED_PREF));
+      hContainer.appendChild(checkbox);
+
+      const viewStudies = this.refs.viewStudies = document.createElementNS(XUL_NS, "label");
+      viewStudies.setAttribute("id", "viewShieldStudies");
+      viewStudies.textContent = "View Firefox Studies";
+      viewStudies.classList.add("learnMore", "text-link");
+      viewStudies.addEventListener("click", this);
```

Comment:
I tried that and got this error:

```
Security Error: Content at moz-nullprincipal:{14b21f4f-1fc8-8046-b8b9-d2b3376239c0} may not load or link to about:studies.
Error: Cannot open a about: link using the text-link binding.
```

Is `about:preferences` running with a null principal, even though it's in the parent process?